### DISCUSSION TITLE
better way to add sequential python and JavaScript events to button?

### DISCUSSION THREAD

**Discussion by retsyo** [zauberzeug/nicegui#2246] [Q&A] [answered]
### Question

In my application, once a button is clicked, a sequence processing should be done in the following order
1. jump to the camere link anchor to show the complete image
this is done in Python code, `ui.open('#cam')`

2. export the page into PDF
this is done by calling JavaScript code, `.props('''onclick="generatePDF()" ''')`

# 1. first try
The code seems simple
```python
ui.button('save', on_click=lambda:ui.open('#cam')).props('''onclick="generatePDF()" ''')
``` 

however, sometimes the exported PDF is blank, I think this happens because navigation to `#cam` has not finished before JavaScript `generatePDF()` is called. 

# 2. second try
So I modified the code into
```python
ui.button('save', on_click=lambda:[ui.open('#cam'), time.sleep(3)]).props('''onclick="generatePDF()" ''')
``` 
but nothing improves. And even it works, `time.sleep(3)` gives a annoying pause.

# 3. third try
Then I edit the code again
```python
async def onSave(e):
        ui.open('#cam')
        time.sleep(3)
        await ui.run_javascript('generatePDF()')

ui.button('存档', on_click=onSave)
```
sometimes, there is only blank image in the exported PDF

# 4. So my question is 
how to revise my code to let JavaScript event `generatePDF()` always happend after Python event `ui.open('#cam')`

Thanks

**Comment by rodja** [comment, marked as answer]
Try

```py
async def onSave(e):
        ui.open('#cam')
        await asyncio.sleep(3)
        ui.run_javascript('generatePDF()')

ui.button('存档', on_click=onSave)
```
